### PR TITLE
Improve Resource API

### LINKED ISSUE #341: ["Request"] Iterable<A>.traverseResource/Iterable<Resource<A>>.sequence
What version are you currently using?

0.11.0/0.12.0-SNAPSHOT

What would you like to see?

`fun <A, B> Iterable<A>.traverse(f: (A) -> Resource<B>): Resource<List<B>>`

### PR DESCRIPTION
**Pull request by i-walker**
## Issue
resolves #341 
and adds 
- traverseResource
- traverseFilterResource
- sequence
- flatTraverseResource
- flatTraverseFilterResource
- flatSequence
- `Resource.Companion.fromClosable`, where `A : java.io.Closable`

## Status
**READY**

## Todos
- [X] Tests
- [X] Documentation

### CODE REVIEW

File: arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/Resource.kt [hunk 1]
```diff
@@ -157,6 +159,9 @@ sealed class Resource<out A> {
   fun <B> zip(other: Resource<B>): Resource<Pair<A, B>> =
     map2(other, ::Pair)
 
+  suspend inline fun <B> foldLeft(inital: B, crossinline f: (B, A) -> B): B =
+    use { f(inital, it) }
```
[the remark this answers is not in the document]
Comment (nomisRev) [reply arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/Resource.kt:163]: That's just `map`

File: arrow-fx-coroutines-test/src/main/kotlin/arrow/fx/coroutines/predef-test.kt [hunk 2]
```diff
@@ -154,6 +154,9 @@ fun <O> Arb.Companion.function(arb: Arb<O>): Arb<() -> O> =
 fun Arb.Companion.unit(): Arb<Unit> =
   Arb.constant(Unit)
 
+fun <A, B> Arb.Companion.functionAToB(arb: Arb<B>): Arb<(A) -> B> = arb.map { b: B -> { _: A -> b } }
+fun <A> Arb.Companion.nullable(arb: Arb<A>): Arb<A?> = Arb.Companion.choice(arb, arb.map { null })
```
Comment (nomisRev) [inline arrow-fx-coroutines-test/src/main/kotlin/arrow/fx/coroutines/predef-test.kt:158]: Small formatting nit 😁

```suggestion
fun <A, B> Arb.Companion.functionAToB(arb: Arb<B>): Arb<(A) -> B> =
  arb.map { b: B -> { _: A -> b } }
  
fun <A> Arb.Companion.nullable(arb: Arb<A>): Arb<A?> =
  Arb.Companion.choice(arb, arb.map { null })
```

File: arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/Resource.kt [hunk 3]
```diff
@@ -159,8 +161,8 @@ sealed class Resource<out A> {
   fun <B> zip(other: Resource<B>): Resource<Pair<A, B>> =
     map2(other, ::Pair)
 
-  suspend inline fun <B> foldLeft(inital: B, crossinline f: (B, A) -> B): B =
-    use { f(inital, it) }
+  fun <B> product(other: Resource<B>): Resource<Tuple2<A, B>> =
```
Comment (nomisRev) [inline arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/Resource.kt]: This is currently duplicated 🤔 but for `Pair` and `Tuple2`.
`product` vs `zip`..

Comment (i-walker) [reply arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/Resource.kt]: Removed product and used zip.

**Review by nomisRev — APPROVED** [review]
Great thanks @i-walker!

File: arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/Resource.kt [hunk 4]
```diff
@@ -372,3 +393,179 @@ infix fun <A> Use<A>.release(release: suspend (A) -> Unit): Resource<A> =
...
+inline fun <A, B> Iterable<A>.traverseResource(crossinline f: (A) -> Resource<B>): Resource<List<B>> =
+  fold(Resource.just(emptyList())) { acc: Resource<List<B>>, a: A ->
+    f(a).ap(acc.map { { b: B -> it + b } })
+  }
+
+/**
+ * Traverses and filters nullable resources
+ * @see traverseResource
+ */
+inline fun <A, B> Iterable<A>.traverseFilterResource(crossinline f: (A) -> Resource<B?>): Resource<List<B>> =
```
Comment (1Jajen1) [inline arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/Resource.kt:442]: Should we add a constraint on `B` to be `B: Any` to not allow users to make the mistake of returning null as a valid value only to have it being filtered out later? Don't think this is super relevant but I think this applies to some other places as well...

**Review by 1Jajen1 — APPROVED** [review]
Not entirely sure about having these combinators, more because they are so specialized to resource and lists, a general solution would be nicer. Other than that this looks good :+1:

**Pull request merged**